ipi/engine/normalmodes.py

```diff
@@ -700,24 +736,27 @@ def get_dynm3(self):
 
     def get_vspring_and_fspring_B(self):
         """
-        Calculates spring forces and potential for bosons.
-        Evaluated using recursion relation from arXiv:1905.090.
+        Calculates spring forces and potential, including exchange effects for the bosons present, if any.
         """
-
         if len(self.bosons) == 0:
-            pass
-        else:
-            (E_k_N, V) = Evaluate_VB(self)
-
-            P = self.nbeads
+            return
 
-            F = np.zeros((P, 3 * self.natoms), float)
-
-            for ind, l in enumerate(self.bosons):
-                for j in range(P):
-                    F[j, 3 * l : 3 * (l + 1)] = Evaluate_dVB(self, E_k_N, V, ind, j)
-
-            return [V[-1], F]
+        masses = dstrip(self.beads.m)[self.bosons]
+        if len(set(masses)) > 1:
+            raise ValueError(
+                "Bosons must have the same mass, found %s for bosons %s"
+                % (str(masses), str(self.bosons))
+            )
+        boson_mass = masses[0]
+        betaP = 1.0 / (self.nbeads * units.Constants.kb * self.ensemble.temp)
+        # positions of only the boson atoms
+        q = dstrip(self.beads.q).reshape((self.nbeads, self.natoms, 3))[
+            :, self.bosons, :
+        ]
+        self.exchange = ExchangePotential(
```

Comment:
I think it'd be better to change the structure of ExchangePotential so you initialize it with the constants (making it perhaps based on depend objects) and then here call it over q.